tools/logstash-docgen/lib/logstash/docgen/parser.rb

```diff
@@ -0,0 +1,143 @@
+# encoding: utf-8
+require "logstash/docgen/dynamic_parser"
+require "logstash/docgen/static_parser"
+require "logstash/docgen/asciidoc_format"
+require "rubygems/specification"
+require "launchy"
+
+module LogStash module Docgen
+  # This class acts as the transformation point between the format
+  # and the data.
+  #
+  # At the beginning of the PoC, I was targetting multiples different format: asciidoc, manpage,
+  # since we only support 1 format now, we could probably remove it.
+  class Document
+    attr_reader :context, :format
+
+    def initialize(context, format)
+      @context = context
+      @format = format
+    end
+
+    def output
+      @output ||= format.generate(context)
+    end
+
+    def save(directory, filename = nil)
+      target_directory = ::File.join(directory, "#{context.section}s")
+      FileUtils.mkdir_p(target_directory)
+
+      filename = "#{context.config_name}.#{format.extension}" unless filename
+      target = ::File.join(target_directory, filename)
+      ::File.open(target, "w") { |f| f.write(output) }
+    end
+  end
+
+  # This class is mostly a data class that represent what will be exposed in the ERB files
+  # Less manipulation should be done in the ERB itself and most of the works should be done here
+  class PluginContext
+    CANONICAL_NAME_PREFIX = "logstash"
+
+    attr_accessor :description, :config_name, :section, :name, :default_plugin, :gemspec
+
+    def initialize(options = {})
+      @config = {}
+      @options = options
+    end
+
+    def default_plugin
+      # HACK: until this is merged https://github.com/elastic/logstash/pull/5096
+      # I have to use some trickery to have the default plugins list.
+      begin
+        path = ::File.expand_path(::File.join(::File.dirname(__FILE__), "..", "..", "..", "..", "..", "rakelib", "default_plugins"))
```

Comment:
Well In an ideal world I would just pull the json and parse it... :(

But #5096 is really close to get merge.